File: sdks/js/samples/BasicPlayerSample/script/app/analytics/video.analytics.provider.js
/*
 * ADOBE SYSTEMS INCORPORATED
 * Copyright 2014 Adobe Systems Incorporated
 * All Rights Reserved.

 * NOTICE:  Adobe permits you to use, modify, and distribute this file in accordance with the
 * terms of the Adobe license agreement accompanying it.  If you have received this file from a
 * source other than Adobe, then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

(function() {
    'use strict';

    var Heartbeat = ADB.va.Heartbeat;
    var HeartbeatConfig = ADB.va.HeartbeatConfig;

    var VideoPlayerPlugin = ADB.va.plugins.videoplayer.VideoPlayerPlugin;
    var VideoPlayerPluginConfig = ADB.va.plugins.videoplayer.VideoPlayerPluginConfig;

    var AdobeAnalyticsPlugin = ADB.va.plugins.aa.AdobeAnalyticsPlugin;
    var AdobeAnalyticsPluginConfig = ADB.va.plugins.aa.AdobeAnalyticsPluginConfig;
    var VideoMetadataKeys = ADB.va.plugins.aa.VideoMetadataKeys;
    var AdMetadataKeys = ADB.va.plugins.aa.AdMetadataKeys;

    var AdobeHeartbeatPlugin = ADB.va.plugins.ah.AdobeHeartbeatPlugin;
    var AdobeHeartbeatPluginConfig = ADB.va.plugins.ah.AdobeHeartbeatPluginConfig;


    function VideoAnalyticsProvider(player) {
        if (!player) {
            throw new Error("Illegal argument. Player reference cannot be null.")
        }
        this._player = player;

        // Set-up the Visitor and AppMeasurement instances.
        var visitor = new Visitor(Configuration.VISITOR.MARKETING_CLOUD_ORG_ID);
        visitor.trackingServer = Configuration.VISITOR.TRACKING_SERVER;
        visitor.setCustomerIDs({
            "userId": {
                "id": Configuration.VISITOR.DPID
            },
            "puuid": {
                "id": Configuration.VISITOR.DPUUID
            }
        });

        // Set-up the AppMeasurement component.
        var appMeasurement = new AppMeasurement();
        appMeasurement.visitor = visitor;
        appMeasurement.trackingServer = Configuration.APP_MEASUREMENT.TRACKING_SERVER;
        appMeasurement.account = Configuration.APP_MEASUREMENT.RSID;
        appMeasurement.pageName = Configuration.APP_MEASUREMENT.PAGE_NAME;
        appMeasurement.charSet = "UTF-8";
        appMeasurement.visitorID = "test-vid";
        appMeasurement.callback = 1;

        // Setup the video-player plugin
        this._playerPlugin = new VideoPlayerPlugin(new SampleVideoPlayerPluginDelegate(this._player));
        var playerPluginConfig = new VideoPlayerPluginConfig();
        playerPluginConfig.debugLogging = true; // set this to false for production apps.
        this._playerPlugin.configure(playerPluginConfig);

        // Setup the AppMeasurement plugin.
        this._aaPlugin = new AdobeAnalyticsPlugin(appMeasurement, new SampleAdobeAnalyticsPluginDelegate());
        var aaPluginConfig = new AdobeAnalyticsPluginConfig();
        aaPluginConfig.channel = Configuration.HEARTBEAT.CHANNEL;
        aaPluginConfig.debugLogging = true; // set this to false for production apps.
        this._aaPlugin.configure(aaPluginConfig);

        // Setup the AdobeHeartbeat plugin.
        var ahPlugin = new AdobeHeartbeatPlugin(new SampleAdobeHeartbeatPluginDelegate());
        var ahPluginConfig = new AdobeHeartbeatPluginConfig(
            Configuration.HEARTBEAT.TRACKING_SERVER,
            Configuration.HEARTBEAT.PUBLISHER);
        ahPluginConfig.ovp = Configuration.HEARTBEAT.OVP;
        ahPluginConfig.sdk = Configuration.HEARTBEAT.SDK;
        ahPluginConfig.debugLogging = true; // set this to false for production apps.
        ahPlugin.configure(ahPluginConfig);

        var plugins = [this._playerPlugin, this._aaPlugin, ahPlugin];

        // Setup and configure the Heartbeat lib.
        this._heartbeat = new Heartbeat(new SampleHeartbeatDelegate(), plugins);
        var configData = new HeartbeatConfig();
        configData.debugLogging = true; // set this to false for production apps.
        this._heartbeat.configure(configData);

        this._installEventListeners();
    }

    VideoAnalyticsProvider.prototype.destroy = function() {
        if (this._player) {
            this._heartbeat.destroy();
            this._heartbeat = null;

            this._uninstallEventListeners();
            this._player = null;
        }
    };


    /////////
    // Notification handlers
    /////////

    VideoAnalyticsProvider.prototype._onLoad = function() {
        console.log('Player event: VIDEO_LOAD');

        var contextData = {};
        // setting Standard Video Metadata
        contextData[VideoMetadataKeys.SEASON] = "sample season";
        contextData[VideoMetadataKeys.SHOW] = "sample show";
        contextData[VideoMetadataKeys.EPISODE] = "sample episode";
        contextData[VideoMetadataKeys.ASSET_ID] = "sample asset id";
        contextData[VideoMetadataKeys.GENRE] = "sample genre";
        contextData[VideoMetadataKeys.FIRST_AIR_DATE] = "sample air date";
        contextData[VideoMetadataKeys.FIRST_DIGITAL_DATE] = "sample digital date";
        contextData[VideoMetadataKeys.RATING] = "sample rating";
        contextData[VideoMetadataKeys.ORIGINATOR] = "sample originator";
        contextData[VideoMetadataKeys.NETWORK] = "sample network";
        contextData[VideoMetadataKeys.SHOW_TYPE] = "sample show type";
        contextData[VideoMetadataKeys.AD_LOAD] = "sample ad load";
        contextData[VideoMetadataKeys.MVPD] = "sample mvpd";
        contextData[VideoMetadataKeys.AUTHORIZED] = "true";
        contextData[VideoMetadataKeys.DAY_PART] = "sample day part";
        contextData[VideoMetadataKeys.FEED] = "sample feed";
        contextData[VideoMetadataKeys.STREAM_FORMAT] = "sample format";

        this._aaPlugin.setVideoMetadata(contextData);
        this._playerPlugin.trackVideoLoad();
    };

    VideoAnalyticsProvider.prototype._onUnload = function() {
        console.log('Player event: VIDEO_UNLOAD');
        this._playerPlugin.trackVideoUnload();
    };

    VideoAnalyticsProvider.prototype._onPlay = function() {
        console.log('Player event: PLAY');
        this._playerPlugin.trackPlay();
    };

    VideoAnalyticsProvider.prototype._onPause = function() {
        console.log('Player event: PAUSE');
        this._playerPlugin.trackPause();
    };

    VideoAnalyticsProvider.prototype._onSeekStart = function() {
        console.log('Player event: SEEK_START');
        this._playerPlugin.trackSeekStart();
    };

    VideoAnalyticsProvider.prototype._onSeekComplete = function() {
        console.log('Player event: SEEK_COMPLETE');
        this._playerPlugin.trackSeekComplete();
    };

    VideoAnalyticsProvider.prototype._onBufferStart = function() {
        console.log('Player event: BUFFER_START');
        this._playerPlugin.trackBufferStart();
    };

    VideoAnalyticsProvider.prototype._onBufferComplete = function() {
        console.log('Player event: BUFFER_COMPLETE');
        this._playerPlugin.trackBufferComplete();
    };

    VideoAnalyticsProvider.prototype._onAdStart = function() {
        console.log('Player event: AD_START');

        var contextData = {};
        // setting Standard Ad Metadata
        contextData[AdMetadataKeys.ADVERTISER] = "sample advertiser";
        contextData[AdMetadataKeys.CAMPAIGN_ID] = "sample campaign";
        contextData[AdMetadataKeys.CREATIVE_ID] = "sample creative";
        contextData[AdMetadataKeys.CREATIVE_URL] = "sample url";
        contextData[AdMetadataKeys.SITE_ID] = "sample site";
        contextData[AdMetadataKeys.PLACEMENT_ID] = "sample placement";

        this._aaPlugin.setAdMetadata(contextData);
        this._playerPlugin.trackAdStart();
    };

    VideoAnalyticsProvider.prototype._onAdComplete = function() {
        console.log('Player event: AD_COMPLETE');
        this._playerPlugin.trackAdComplete();
    };

    VideoAnalyticsProvider.prototype._onChapterStart = function() {
        console.log('Player event: CHAPTER_START');
        this._aaPlugin.setChapterMetadata({
            segmentType: "Sample segment type"
        });
        this._playerPlugin.trackChapterStart();
    };

    VideoAnalyticsProvider.prototype._onChapterComplete = function() {
        console.log('Player event: CHAPTER_COMPLETE');
        this._playerPlugin.trackChapterComplete();
    };

    VideoAnalyticsProvider.prototype._onComplete = function() {
        console.log('Player event: COMPLETE');
        this._playerPlugin.trackComplete(function() {
            console.log("The completion of the content has been tracked.");
        });
    };


    /////////
    // Private helper functions
    /////////

    VideoAnalyticsProvider.prototype._installEventListeners = function() {
        // We register as observers to various VideoPlayer events.
        NotificationCenter().addEventListener(PlayerEvent.VIDEO_LOAD, this._onLoad, this);
        NotificationCenter().addEventListener(PlayerEvent.VIDEO_UNLOAD, this._onUnload, this);
        NotificationCenter().addEventListener(PlayerEvent.PLAY, this._onPlay, this);
        NotificationCenter().addEventListener(PlayerEvent.PAUSE, this._onPause, this);
        NotificationCenter().addEventListener(PlayerEvent.SEEK_START, this._onSeekStart, this);
        NotificationCenter().addEventListener(PlayerEvent.SEEK_COMPLETE, this._onSeekComplete, this);
        NotificationCenter().addEventListener(PlayerEvent.BUFFER_START, this._onBufferStart, this);
        NotificationCenter().addEventListener(PlayerEvent.BUFFER_COMPLETE, this._onBufferComplete, this);
        NotificationCenter().addEventListener(PlayerEvent.AD_START, this._onAdStart, this);
        NotificationCenter().addEventListener(PlayerEvent.AD_COMPLETE, this._onAdComplete, this);
        NotificationCenter().addEventListener(PlayerEvent.CHAPTER_START, this._onChapterStart, this);
        NotificationCenter().addEventListener(PlayerEvent.CHAPTER_COMPLETE, this._onChapterComplete, this);
        NotificationCenter().addEventListener(PlayerEvent.COMPLETE, this._onComplete, this);
    };

    VideoAnalyticsProvider.prototype._uninstallEventListeners = function() {
        // We register as observers to various VideoPlayer events.
        NotificationCenter().removeEventListener(PlayerEvent.VIDEO_LOAD, this._onLoad, this);
        NotificationCenter().removeEventListener(PlayerEvent.VIDEO_UNLOAD, this._onUnload, this);
        NotificationCenter().removeEventListener(PlayerEvent.PLAY, this._onPlay, this);
        NotificationCenter().removeEventListener(PlayerEvent.PAUSE, this._onPause, this);
        NotificationCenter().removeEventListener(PlayerEvent.SEEK_START, this._onSeekStart, this);
        NotificationCenter().removeEventListener(PlayerEvent.SEEK_COMPLETE, this._onSeekComplete, this);
        NotificationCenter().removeEventListener(PlayerEvent.BUFFER_START, this._onBufferStart, this);
        NotificationCenter().removeEventListener(PlayerEvent.BUFFER_COMPLETE, this._onBufferComplete, this);
        NotificationCenter().removeEventListener(PlayerEvent.AD_START, this._onAdStart, this);
        NotificationCenter().removeEventListener(PlayerEvent.AD_COMPLETE, this._onAdComplete, this);
        NotificationCenter().removeEventListener(PlayerEvent.CHAPTER_START, this._onChapterStart, this);
        NotificationCenter().removeEventListener(PlayerEvent.CHAPTER_COMPLETE, this._onChapterComplete, this);
        NotificationCenter().removeEventListener(PlayerEvent.COMPLETE, this._onComplete, this);
    };

    // Export symbols.
    window.VideoAnalyticsProvider = VideoAnalyticsProvider;
})();

File: sdks/android/samples/BasicPlayerSample/src/com/adobe/primetime/va/samples/analytics/VideoAnalyticsProvider.java
/*
 * ADOBE SYSTEMS INCORPORATED
 * Copyright 2014 Adobe Systems Incorporated
 * All Rights Reserved.

 * NOTICE:  Adobe permits you to use, modify, and distribute this file in accordance with the
 * terms of the Adobe license agreement accompanying it.  If you have received this file from a
 * source other than Adobe, then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

package com.adobe.primetime.va.samples.analytics;

import android.util.Log;

import com.adobe.mobile.Config;
import com.adobe.primetime.core.ICallback;
import com.adobe.primetime.core.plugin.IPlugin;
import com.adobe.primetime.va.Heartbeat;
import com.adobe.primetime.va.plugins.aa.AdMetadataKeys;
import com.adobe.primetime.va.plugins.aa.AdobeAnalyticsPlugin;
import com.adobe.primetime.va.plugins.aa.AdobeAnalyticsPluginConfig;
import com.adobe.primetime.va.plugins.aa.VideoMetadataKeys;
import com.adobe.primetime.va.plugins.ah.AdobeHeartbeatPluginConfig;
import com.adobe.primetime.va.plugins.videoplayer.VideoPlayerPlugin;
import com.adobe.primetime.va.plugins.ah.AdobeHeartbeatPlugin;
import com.adobe.primetime.va.plugins.videoplayer.VideoPlayerPluginConfig;
import com.adobe.primetime.va.samples.Configuration;
import com.adobe.primetime.va.samples.player.PlayerEvent;
import com.adobe.primetime.va.samples.player.VideoPlayer;

import java.util.*;

public class VideoAnalyticsProvider implements Observer {
    private static final String LOG_TAG = "[VideoHeartbeatSample]::" + VideoAnalyticsProvider.class.getSimpleName();

    private VideoPlayer _player;
    private VideoPlayerPlugin _playerPlugin;
    private AdobeAnalyticsPlugin _aaPlugin;
    private AdobeHeartbeatPlugin _ahPlugin;
    private Heartbeat _heartbeat;

    public VideoAnalyticsProvider(VideoPlayer player) {
        if (player == null) {
            throw new IllegalArgumentException("Player reference cannot be null.");
        }
        _player = player;
        _player.addObserver(this);

        // Set the plugin list.
        List<IPlugin> plugins = new ArrayList<IPlugin>();

        // Setup VideoPlayer plugin
        _playerPlugin = new VideoPlayerPlugin(new SampleVideoPlayerPluginDelegate(_player));
        VideoPlayerPluginConfig playerPluginConfig = new VideoPlayerPluginConfig();
        playerPluginConfig.debugLogging = true; // set this to false for production apps.
        _playerPlugin.configure(playerPluginConfig);
        plugins.add(_playerPlugin);

        // Setup the visitor id - optional parameter to be set by the player.
        Config.setUserIdentifier("test-vid");

        // Setup AdobeAnalytics plugin
        _aaPlugin = new AdobeAnalyticsPlugin(new SampleAdobeAnalyticsPluginDelegate());
        AdobeAnalyticsPluginConfig aaPluginConfig = new AdobeAnalyticsPluginConfig();
        aaPluginConfig.channel = Configuration.HEARTBEAT_CHANNEL;
        aaPluginConfig.debugLogging = true; // set this to false for production apps.
        _aaPlugin.configure(aaPluginConfig);
        plugins.add(_aaPlugin);

        // Setup AdobeHeartbeat plugin
        _ahPlugin = new AdobeHeartbeatPlugin(new SampleAdobeHeartbeatPluginDelegate());
        AdobeHeartbeatPluginConfig ahPluginConfig = new AdobeHeartbeatPluginConfig(
                Configuration.HEARTBEAT_TRACKING_SERVER,
                Configuration.HEARTBEAT_PUBLISHER
        );
        ahPluginConfig.ovp = Configuration.HEARTBEAT_OVP;
        ahPluginConfig.sdk = Configuration.HEARTBEAT_SDK;
        ahPluginConfig.debugLogging = true; // set this to false for production apps.
        _ahPlugin.configure(ahPluginConfig);
        plugins.add(_ahPlugin);

        // Setup and configure Heartbeat lib
        _heartbeat = new Heartbeat(new SampleHeartbeatDelegate(), plugins);
    }

    public void destroy() {
        if (_player != null) {
            _heartbeat.destroy();
            _heartbeat = null;
            _aaPlugin = null;
            _playerPlugin = null;
            _ahPlugin = null;

            _player.destroy();
            _player.deleteObserver(this);
            _player = null;
        }
    }

    @Override
    public void update(Observable observable, Object data) {
        PlayerEvent playerEvent = (PlayerEvent) data;

        switch (playerEvent) {
            case VIDEO_LOAD:
                Log.d(LOG_TAG, "Video loaded.");
                HashMap<String, String> videoMetadata = new HashMap<String, String>();
                videoMetadata.put(VideoMetadataKeys.EPISODE, "Sample Episode");
                videoMetadata.put(VideoMetadataKeys.SEASON, "Sample Season");
                videoMetadata.put(VideoMetadataKeys.SHOW, "Sample Show");
                videoMetadata.put(VideoMetadataKeys.ASSET_ID, "Sample Asset ID");
                videoMetadata.put(VideoMetadataKeys.GENRE, "Sample Genre");
                videoMetadata.put(VideoMetadataKeys.FIRST_AIR_DATE, "Sample air date");
                videoMetadata.put(VideoMetadataKeys.FIRST_DIGITAL_DATE, "Sample digital date");
                videoMetadata.put(VideoMetadataKeys.RATING, "Sample Rating");
                videoMetadata.put(VideoMetadataKeys.ORIGINATOR, "Sample Originator");
                videoMetadata.put(VideoMetadataKeys.SHOW_TYPE, "Sample Show Type");
                videoMetadata.put(VideoMetadataKeys.AD_LOAD, "ad load");
                videoMetadata.put(VideoMetadataKeys.MVPD, "sample mvpd");
                videoMetadata.put(VideoMetadataKeys.AUTHORIZED, "false");
                videoMetadata.put(VideoMetadataKeys.NETWORK, "Sample TV Network");
                videoMetadata.put(VideoMetadataKeys.DAY_PART, "sample day part");
                videoMetadata.put(VideoMetadataKeys.FEED, "sample feed type");
                videoMetadata.put(VideoMetadataKeys.STREAM_FORMAT, "sample stream format");
                _aaPlugin.setVideoMetadata(videoMetadata);

                _playerPlugin.trackVideoLoad();
                break;

            case VIDEO_UNLOAD:
                Log.d(LOG_TAG, "Video unloaded.");
                _playerPlugin.trackVideoUnload();
                break;

            case PLAY:
                Log.d(LOG_TAG, "Playback started.");
                _playerPlugin.trackPlay();
                break;

            case PAUSE:
                Log.d(LOG_TAG, "Playback paused.");
                _playerPlugin.trackPause();
                break;

            case SEEK_START:
                Log.d(LOG_TAG, "Seek started.");
                _playerPlugin.trackSeekStart();
                break;

            case SEEK_COMPLETE:
                Log.d(LOG_TAG, "Seek completed.");
                _playerPlugin.trackSeekComplete();
                break;

            case BUFFER_START:
                Log.d(LOG_TAG, "Buffer started.");
                _playerPlugin.trackBufferStart();
                break;

            case BUFFER_COMPLETE:
                Log.d(LOG_TAG, "Buffer completed.");
                _playerPlugin.trackBufferComplete();
                break;

            case AD_START:
                Log.d(LOG_TAG, "Ad started.");
                HashMap<String, String> adMetadata = new HashMap<String, String>();
                adMetadata.put(AdMetadataKeys.CREATIVE_ID, "Sample creative id");
                adMetadata.put(AdMetadataKeys.CAMPAIGN_ID, "Sample ad campaign");
                adMetadata.put(AdMetadataKeys.ADVERTISER, "Sample advertiser");
                adMetadata.put(AdMetadataKeys.CREATIVE_URL, "Sample creative url");
                adMetadata.put(AdMetadataKeys.PLACEMENT_ID, "Sample placement");
                adMetadata.put(AdMetadataKeys.SITE_ID, "Sample site");
                _aaPlugin.setAdMetadata(adMetadata);

                _playerPlugin.trackAdStart();
                break;

            case AD_COMPLETE:
                Log.d(LOG_TAG, "Ad completed.");
                _playerPlugin.trackAdComplete();
                break;

            case CHAPTER_START:
                Log.d(LOG_TAG, "Chapter started.");
                HashMap<String, String> chapterMetadata = new HashMap<String, String>();
                chapterMetadata.put("segmentType", "Sample Segment Type");
                _aaPlugin.setChapterMetadata(chapterMetadata);

                _playerPlugin.trackChapterStart();
                break;

            case CHAPTER_COMPLETE:
                Log.d(LOG_TAG, "Chapter completed.");
                _playerPlugin.trackChapterComplete();
                break;

            case COMPLETE:
                Log.d(LOG_TAG, "Playback completed.");

                ICallback completeCallback = new ICallback() {
                    @Override
                    public Object call(Object param) {
                        Log.d(LOG_TAG, "The completion of the content has been tracked.");
                        return null;
                    }
                };

                _playerPlugin.trackComplete(completeCallback);
                break;

            default:
                Log.d(LOG_TAG, "Unhandled player event: " + playerEvent.toString());
                break;
        }
    }
}
